File: adapters/src/main/java/com/wang/adapters/BaseContainerAdapter.java
package com.wang.adapters;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.collection.SimpleArrayMap;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.wang.adapters.adapter.BaseContainerItemAdapter;
import com.wang.adapters.adapter.IContainerItemAdapter;
import com.wang.adapters.bean.IContainerBean;
import com.wang.adapters.observer.IContainerObserver;

import java.util.ArrayList;
import java.util.List;

/**
 * 一个超级adapter可以添加其他adapter
 * <p>
 * 可以用在如：天猫首页、bilibili、今日头条、聊天列表页面
 * <p>
 * 核心思想：每个{@link BEAN}的item都当作一个adapter，为了复用和管理相同adapter，在子adapter传入了{@link BEAN}{@link IContainerItemAdapter#getCurrentBean}
 * <p>
 * 限制条件：
 * 1.bean必须继承{@link IContainerBean}
 * 2.子adapter必须是{@link IContainerItemAdapter},{@link BaseContainerItemAdapter}的子类
 * <p>
 * 3.子adapter的type必须在±10000之间{@link #TYPE_MAX}{@link #TYPE_MIN}
 * 4.如果是GridLayoutManager必须提前设置（在rv.setAdapter或addAdapter之前或手动调用）
 * 其他限制暂时没发现
 */
@SuppressWarnings({"WeakerAccess", "unused"})
public final class BaseContainerAdapter<BEAN extends IContainerBean> extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    protected final String TAG = getClass().getSimpleName();

    public static final int TYPE_MAX = 100000, TYPE_MIN = -100000;
    private static final int TYPE_MINUS = TYPE_MAX - TYPE_MIN;

    protected MyItemAdapter mItemAdapters = new MyItemAdapter();
    protected IContainerObserver mObservers = new IContainerObserver() {
        @Override
        public void notifyDataSetChanged() {
            BaseContainerAdapter.this.notifyDataSetChanged();
        }

        @Override
        public void notifyItemChanged(int position, IContainerBean bean) {
            notifyItemChanged(position, 1, bean);
        }

        @Override
        public void notifyItemChanged(int positionStart, int itemCount, IContainerBean bean) {
            int newPosition = getBaseAdapterPosition(bean, positionStart);
            BaseContainerAdapter.this.notifyItemRangeChanged(newPosition, itemCount, null);
        }

        @Override
        public void notifyItemInserted(int position, IContainerBean bean) {
            notifyItemInserted(position, 1, bean);
        }

        @Override
        public void notifyItemInserted(int positionStart, int itemCount, IContainerBean bean) {
            int newPosition = getBaseAdapterPosition(bean, positionStart);
            BaseContainerAdapter.this.notifyItemRangeInserted(newPosition, itemCount);
        }

        @Override
        public void notifyItemMoved(int fromPosition, int toPosition, IContainerBean bean) {
            int newPosition = getBaseAdapterPosition(bean, fromPosition);
            BaseContainerAdapter.this.notifyItemMoved(newPosition, newPosition + (toPosition - fromPosition));
        }

        @Override
        public void notifyItemRemoved(int position, IContainerBean bean) {
            notifyItemRemoved(position, 1, bean);
        }

        @Override
        public void notifyItemRemoved(int positionStart, int itemCount, IContainerBean bean) {
            int newPosition = getBaseAdapterPosition(bean, positionStart);
            BaseContainerAdapter.this.notifyItemRangeRemoved(newPosition, itemCount);
        }
    };

    protected RecyclerView mRecyclerView;
    protected GridLayoutManager mLayoutManager;

    @NonNull
    private List<BEAN> mList;

    /**
     * 子adapter的信息,具体见{@link MyItemInfo#refreshItemInfo}
     */
    protected MyItemInfo mItemInfo = new MyItemInfo();

    /**
     * 注释同下
     */
    public BaseContainerAdapter() {
        this(null);
    }

    /**
     * 注意：如果是GridLayoutManager请提前设置好（在rv.setAdapter或addAdapter之前或手动调用）
     * 如果中途更换LayoutManager请调用{@link #changedLayoutManager}
     */
    public BaseContainerAdapter(@Nullable List<BEAN> list) {
        mList = list == null ? new ArrayList<BEAN>() : list;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return mItemAdapters.getAdapter(viewType / TYPE_MINUS).createViewHolder(parent, viewType % TYPE_MINUS, LayoutInflater.from(parent.getContext()));
    }

    @SuppressWarnings("unchecked")//未检查警告,此处忽略
    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        mItemInfo.refreshItemInfo(position);
        IContainerItemAdapter itemAdapter = mItemInfo.mItemAdapter;
        itemAdapter.setCurrentBean(mList.get(mItemInfo.mListPosition));
        itemAdapter.bindViewHolder(holder, mItemInfo.mItemPosition);
    }

    @SuppressWarnings("unchecked")//未检查警告,此处忽略
    @Override
    public int getItemViewType(int position) {
        mItemInfo.refreshItemInfo(position);
        IContainerItemAdapter itemAdapter = mItemInfo.mItemAdapter;
        itemAdapter.setCurrentBean(mList.get(mItemInfo.mListPosition));
        int itemType = itemAdapter.getItemViewType(mItemInfo.mItemPosition);
        if (itemType < TYPE_MIN || itemType >= TYPE_MAX) {
            throw new RuntimeException("你的adapter" + itemAdapter.getClass() + "的type必须在" + TYPE_MIN + "~" + TYPE_MAX + "之间，");
        }
        //根据mItemAdapters的position返回type，取的时候比较方便
        return mItemAdapters.getPosition(itemAdapter.getClass()) * TYPE_MINUS + itemType;
    }

    @SuppressWarnings("unchecked")//未检查警告,此处忽略
    @Override
    public int getItemCount() {
        int count = 0;
        for (BEAN bean : mList) {
            IContainerItemAdapter itemAdapter = mItemAdapters.getAdapter(bean.getItemAdapterClass());
            itemAdapter.setCurrentBean(bean);
            count += itemAdapter.getItemCount();
        }
        return count;
    }

    @Override
    public void onAttachedToRecyclerView(@NonNull RecyclerView recyclerView) {
        mRecyclerView = recyclerView;
        checkLayoutManager();
    }

    protected void checkLayoutManager() {
        if (mRecyclerView == null) return;
        RecyclerView.LayoutManager manager = mRecyclerView.getLayoutManager();
        if ((mLayoutManager == null || mLayoutManager != manager) && manager instanceof GridLayoutManager) {
            changedLayoutManager((GridLayoutManager) manager);
        }
    }

    /**
     * 根据bean对象和adapter的相对位置获取绝对位置
     */
    @SuppressWarnings("unchecked")//未检查警告,此处忽略
    protected int getBaseAdapterPosition(IContainerBean bean, int itemAdapterPosition) {
        int position = itemAdapterPosition;
        for (BEAN listBean : mList) {
            if (listBean == bean) {
                return position;
            } else {
                IContainerItemAdapter itemAdapter = mItemAdapters.getAdapter(bean.getItemAdapterClass());
                itemAdapter.setCurrentBean(listBean);
                position += itemAdapter.getItemCount();
            }
        }
        throw new RuntimeException("在list中没有找到传入的bean对象" + bean);
    }

    protected class MyItemInfo {
        /**
         * 使用之前请调用{@link #refreshItemInfo}
         * list的position,子adapter的所需要的相对position
         */
        protected int mListPosition, mItemPosition;
        protected IContainerItemAdapter mItemAdapter;

        /**
         * 根据超级adapter的position返回子adapter的信息
         */
        @SuppressWarnings("unchecked")//未检查警告,此处忽略
        protected MyItemInfo refreshItemInfo(int position) {
            //itemAdapter的position=0时的真实位置
            int itemStartPosition = 0;
            for (int i = 0; i < mList.size(); i++) {
                BEAN bean = mList.get(i);
                IContainerItemAdapter itemAdapter = mItemAdapters.getAdapter(bean.getItemAdapterClass());
                itemAdapter.setCurrentBean(bean);
                int itemCount = itemAdapter.getItemCount();
                int nextStartPosition = itemStartPosition + itemCount;
                //下一个adapter的位置比position大说明当前type就在这个adapter中
                if (nextStartPosition > position) {
                    mListPosition = i;
                    mItemAdapter = itemAdapter;
                    mItemPosition = position - itemStartPosition;
                    return this;
                } else {
                    //循环相加
                    itemStartPosition = nextStartPosition;
                }
            }
            throw new RuntimeException("没有取到对应的type,可能你没有(及时)刷新adapter");
        }
    }

    /**
     * position,adapter,class唯一并且可以互相取值
     */
    protected class MyItemAdapter {
        protected SimpleArrayMap<Class<? extends IContainerItemAdapter>, Integer> mMap = new SimpleArrayMap<>(8);
        protected ArrayList<IContainerItemAdapter> mList = new ArrayList<>(8);

        protected void addAdapter(IContainerItemAdapter... adapters) {
            for (IContainerItemAdapter adapter : adapters) {
                adapter.registerDataSetObserver(mObservers);
                if (!mMap.containsKey(adapter.getClass())) {
                    mList.add(adapter);
                    mMap.put(adapter.getClass(), mList.size() - 1);
                }
            }
        }

        protected void addAdapter(List<? extends IContainerItemAdapter> adapters) {
            for (IContainerItemAdapter adapter : adapters) {
                adapter.registerDataSetObserver(mObservers);
                if (!mMap.containsKey(adapter.getClass())) {
                    mList.add(adapter);
                    mMap.put(adapter.getClass(), mList.size() - 1);
                }
            }
        }

        /**
         * @throws RuntimeException 见{@link #getPosition}
         */
        protected IContainerItemAdapter getAdapter(int position) {
            if (position < 0 || position >= mList.size()) {
                throw new RuntimeException("缺少对应的adapter，adapter数量：" + mList.size() + "，当前数量：" + position);
            }
            return mList.get(position);
        }

        /**
         * @throws RuntimeException 见{@link #getPosition}
         */
        protected IContainerItemAdapter getAdapter(Class<? extends IContainerItemAdapter> cls) {
            Integer index = mMap.get(cls);
            if (index == null) {
                //
                throw new RuntimeException("缺少对应的adapter：" + cls);
            }
            return mList.get(index);
        }

        /**
         * @throws NullPointerException 一般是数据变化没有(及时)刷新adapter导致的
         */
        @SuppressWarnings("ConstantConditions")
        protected int getPosition(Class<? extends IContainerItemAdapter> cls) {
            return mMap.get(cls);
        }

        protected void remove(int position) {
            IContainerItemAdapter remove = mList.remove(position);
            mMap.remove(remove.getClass());
        }

        protected void remove(Class<? extends IContainerItemAdapter> cls) {
            Integer remove = mMap.remove(cls);
            if (remove != null) {
                mList.remove((int) remove);
            }
        }

        protected void clear() {
            mList.clear();
            mMap.clear();
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    // 以下是可调用方法
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    /**
     * 添加adapter.重复则不会被添加,必须先删除
     * 当然可以预先添加用不到的adapter
     */
    public BaseContainerAdapter<BEAN> addAdapter(@NonNull IContainerItemAdapter... adapters) {
        mItemAdapters.addAdapter(adapters);
        checkLayoutManager();
        notifyDataSetChanged();
        return this;
    }

    /**
     * 添加adapter.重复则不会被添加,必须先删除
     * 当然可以预先添加用不到的adapter
     */
    public BaseContainerAdapter<BEAN> addAdapter(@NonNull List<? extends IContainerItemAdapter<? extends RecyclerView.ViewHolder, BEAN>> adapters) {
        mItemAdapters.addAdapter(adapters);
        checkLayoutManager();
        notifyDataSetChanged();
        return this;
    }

    /**
     * 设置新的list数据并刷新adapter
     */
    public void setListAndNotifyDataSetChanged(List<BEAN> list) {
        if (list == null) {
            mList.clear();
        } else {
            mList = list;
        }
        notifyDataSetChanged();
    }

    @NonNull
    public List<BEAN> getList() {
        return mList;
    }

    /**
     * 删除指定adapter
     * 别忘了同步修改list数据和{@link #notifyDataSetChanged}
     *
     * @param adapterPosition 按添加顺序第几个
     */
    public BaseContainerAdapter<BEAN> removeAdapter(int adapterPosition) {
        mItemAdapters.remove(adapterPosition);
        return this;
    }

    /**
     * 删除指定adapter
     * 别忘了同步修改list数据和{@link #notifyDataSetChanged}
     *
     * @param adapterClass 哪个adapter
     */
    public BaseContainerAdapter<BEAN> removeAdapter(Class<? extends IContainerItemAdapter> adapterClass) {
        mItemAdapters.remove(adapterClass);
        return this;
    }

    /**
     * 清空adapter（数据不同步清空会崩溃哦）
     * 别忘了同步修改list数据和{@link #notifyDataSetChanged}
     */
    public BaseContainerAdapter<BEAN> removeAllAdapter() {
        mItemAdapters.clear();
        return this;
    }

    /**
     * 把rv的LayoutManager改成其他的GridLayoutManager时.此方法理论上没啥用
     */
    @SuppressWarnings("unchecked")//未检查警告,此处忽略
    public void changedLayoutManager(GridLayoutManager manager) {
        if (manager == null) return;
        mLayoutManager = manager;
        mLayoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
            @Override
            public int getSpanSize(int position) {
                mItemInfo.refreshItemInfo(position);
                IContainerItemAdapter itemAdapter = mItemInfo.mItemAdapter;
                itemAdapter.setCurrentBean(mList.get(mItemInfo.mListPosition));
                return itemAdapter.getSpanSize(mItemInfo.mItemPosition);
            }
        });
    }
}

File: adapters/src/main/java/com/wang/container/BaseContainerAdapter.kt
package com.wang.container

import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.view.View
import android.view.ViewGroup
import androidx.annotation.MainThread
import androidx.collection.SimpleArrayMap
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.GridLayoutManager.SpanSizeLookup
import androidx.recyclerview.widget.RecyclerView
import androidx.viewbinding.ViewBinding
import com.wang.container.BaseContainerAdapter.Companion.TYPE_MAX
import com.wang.container.BaseContainerAdapter.Companion.TYPE_MIN
import com.wang.container.adapter.BaseContainerItemAdapter
import com.wang.container.adapter.OneContainerItemAdapter
import com.wang.container.bean.IContainerBean
import com.wang.container.bean.ItemAdapterPositionInfo
import com.wang.container.helper.BaseListAdapterHelper
import com.wang.container.holder.BaseViewHolder
import com.wang.container.interfaces.DefOnItemClickListener
import com.wang.container.interfaces.IListAdapter
import com.wang.container.interfaces.OnItemClickListener
import com.wang.container.observer.IContainerObserver
import com.wang.container.utils.castSuperAdapter

/**
 * 一个超级adapter可以添加其他adapter
 *
 *
 * 可以用在如：天猫首页、bilibili、今日头条、聊天列表页面
 *
 *
 * 核心思想：每个[BEAN]的item都当作一个adapter，所以再调用时都有个currentBean，adapter更新/判断数据时以currentBean为准
 *
 *
 * 使用前提（都是无关紧要的，但也要看看）：
 * 1.bean必须继承[IContainerBean]
 * 2.子adapter必须是[BaseContainerItemAdapter]、[OneContainerItemAdapter]的子类
 * 3.子adapter的type必须在[TYPE_MAX]、[TYPE_MIN]之间
 * 4.如果是GridLayoutManager必须在adapter前设置（在rv.setAdapter或[addAdapter]之前或手动调用[changedLayoutManager]）
 * 5.有header时直接调用BaseContainerAdapter的[notifyItemChanged]相关方法时需要+1（所有adapter的通病，建议使用[notifyListItemChanged]）（子adapter刷新时无需考虑父的header）
 * 其他限制暂时没发现
 *
 *
 * https://blog.csdn.net/weimingjue/article/details/106468916
 */
@SuppressLint("NotifyDataSetChanged")
class BaseContainerAdapter<BEAN : IContainerBean> @JvmOverloads constructor(list: List<BEAN>? = null) :
    RecyclerView.Adapter<BaseViewHolder<*>>(),
    IListAdapter<BEAN, ViewBinding, OnItemClickListener<BEAN>> {

    companion object {
        const val TYPE_MAX = 100_000
        const val TYPE_MIN = -100_000

        /**
         * 内部返回的type永远>=[TYPE_MIN]，所以不会和adapter重复
         */
        const val TYPE_HEADER = TYPE_MIN - 1
        const val TYPE_FOOTER = TYPE_MIN - 2
        const val TYPE_MINUS = TYPE_MAX - TYPE_MIN
    }

    private var lastCachePositionInfo: ItemAdapterPositionInfo? = null
    private var internalLastCachePositionInfo: ItemAdapterPositionInfo? = null
    private var onItemClickListener: OnItemClickListener<BEAN>? = null
    private val adaptersManager = MyAdaptersManager()
    private val childObservers: IContainerObserver = object : IContainerObserver {
        override fun notifyDataSetChanged() {
            this@BaseContainerAdapter.notifyDataSetChanged()
        }

        override fun notifyItemChanged(
            relativePositionStart: Int,
            itemCount: Int,
            bean: IContainerBean
        ) {
            val newPosition = getAbsPosition(bean, relativePositionStart)
            this@BaseContainerAdapter.notifyItemRangeChanged(newPosition, itemCount)
        }

        override fun notifyItemInserted(
            relativePositionStart: Int,
            itemCount: Int,
            bean: IContainerBean
        ) {
            val newPosition = getAbsPosition(bean, relativePositionStart)
            this@BaseContainerAdapter.notifyItemRangeInserted(newPosition, itemCount)
        }

        override fun notifyItemMoved(
            relativeFromPosition: Int,
            relativePositionToPosition: Int,
            bean: IContainerBean
        ) {
            val newPosition = getAbsPosition(bean, relativeFromPosition)
            this@BaseContainerAdapter.notifyItemMoved(
                newPosition,
                newPosition + (relativePositionToPosition - relativeFromPosition)
            )
        }

        override fun notifyItemRemoved(
            relativePositionStart: Int,
            itemCount: Int,
            bean: IContainerBean
        ) {
            val newPosition = getAbsPosition(bean, relativePositionStart)
            this@BaseContainerAdapter.notifyItemRangeRemoved(newPosition, itemCount)
        }

        override fun dispatchItemClicked(view: View) {
            onItemClickListener?.onClick(view)
        }

        override fun dispatchItemLongClicked(view: View): Boolean {
            return onItemClickListener?.onLongClick(view) ?: false
        }
    }
    private var recyclerView: RecyclerView? = null
    private var lastLayoutManager: GridLayoutManager? = null

    /**
     * list相关代码合并
     */
    private val listHelper = BaseListAdapterHelper(this, list)

    init {
        registerAdapterDataObserver(object : RecyclerView.AdapterDataObserver() {
            override fun onChanged() {
                super.onChanged()
                onAdapterChanged()
            }

            override fun onItemRangeChanged(positionStart: Int, itemCount: Int) {
                super.onItemRangeChanged(positionStart, itemCount)
                onAdapterChanged()
            }

            override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
                super.onItemRangeInserted(positionStart, itemCount)
                onAdapterChanged()
            }

            override fun onItemRangeRemoved(positionStart: Int, itemCount: Int) {
                super.onItemRangeRemoved(positionStart, itemCount)
                onAdapterChanged()
            }

            override fun onItemRangeMoved(fromPosition: Int, toPosition: Int, itemCount: Int) {
                super.onItemRangeMoved(fromPosition, toPosition, itemCount)
                onAdapterChanged()
            }

            fun onAdapterChanged() {
                lastCachePositionInfo = null
                internalLastCachePositionInfo?.absListPosition = -999
            }
        })
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder<*> {
        return when (viewType) {
            TYPE_HEADER, TYPE_FOOTER -> {
                listHelper.onCreateHeaderFooterViewHolder(parent)
            }
            else -> {
                adaptersManager.getAdapter(viewType / TYPE_MINUS)
                    .createViewHolder(parent, viewType % TYPE_MINUS)
            }
        }
    }

    override fun onBindViewHolder(holder: BaseViewHolder<*>, position: Int) {
        when (getItemViewType(position)) {
            TYPE_HEADER -> {
                listHelper.onBindHeaderFooterViewHolder(holder, listHelper.headerView!!)
            }
            TYPE_FOOTER -> {
                listHelper.onBindHeaderFooterViewHolder(holder, listHelper.footerView!!)
            }
            else -> {
                val info = getCacheItemPositionInfo(position, true)
                val itemAdapter = info.itemAdapter.castSuperAdapter()
                itemAdapter.bindViewHolder(
                    holder,
                    list[info.absListPosition],
                    info.itemRelativePosition
                )
            }
        }
    }

    @Deprecated("暂时无法实现，请勿调用", ReplaceWith("无"))
    @TargetApi(999)
    override fun onCreateListViewHolder(parent: ViewGroup): BaseViewHolder<ViewBinding> {
        //这两个方法是给其他adapter用的：https://github.com/weimingjue/BaseAdapter
        throw RuntimeException("暂时无法实现，请勿调用")
    }

    @Deprecated("暂时无法实现，请勿调用", ReplaceWith("无"))
    @TargetApi(999)
    override fun onBindListViewHolder(
        holder: BaseViewHolder<ViewBinding>,
        listPosition: Int,
        bean: BEAN
    ) {
        throw RuntimeException("暂时无法实现，请勿调用")
    }

    override fun getItemViewType(position: Int): Int {
        if (isHeaderView && position == 0) {
            return TYPE_HEADER
        }
        if (isFooterView && itemCount == position + 1) {
            return TYPE_FOOTER
        }
        val info = getCacheItemPositionInfo(position, true)
        val itemAdapter = info.itemAdapter.castSuperAdapter()
        val itemType =
            itemAdapter.getItemViewType(list[info.absListPosition], info.itemRelativePosition)
        if (itemType < TYPE_MIN || itemType >= TYPE_MAX) {
            throw RuntimeException("你adapter（" + itemAdapter.javaClass + "）的type必须在" + TYPE_MIN + "~" + TYPE_MAX + "之间，type：" + itemType)
        }
        //根据mItemAdapters的position返回type，取的时候比较方便
        //此处返回的type>=TYPE_MIN
        return adaptersManager.getPosition(itemAdapter.javaClass) * TYPE_MINUS + itemType
    }

    override fun getItemCount(): Int {
        var count = headerViewCount + footerViewCount
        listHelper.list.forEach { bean ->
            val itemAdapter =
                adaptersManager.getAdapter(bean.getBindAdapterClass()).castSuperAdapter()
            count += itemAdapter.getItemCount(bean)
        }
        return count
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        this.recyclerView = recyclerView
        checkLayoutManager()
    }

    private fun checkLayoutManager() {
        (recyclerView?.layoutManager as? GridLayoutManager)?.let {
            if (lastLayoutManager != it) {
                changedLayoutManager(it)
            }
        }
    }

    /**
     * 根据绝对position获取子adapter的相关信息
     *
     * @param absPosition 绝对position
     * @param internalRecycle 开启内部循环利用，返回值绝对不可声明final（当然为了安全，外部调用默认false）
     */
    internal fun getCacheItemPositionInfo(
        absPosition: Int,
        internalRecycle: Boolean
    ): ItemAdapterPositionInfo {
        val absListPosition = absPosition - headerViewCount

        //取缓存
        val cache = if (internalRecycle) internalLastCachePositionInfo else lastCachePositionInfo
        if (cache?.absListPosition == absListPosition) {
            return cache
        }

        //itemAdapter的position=0时的真实位置
        var itemStartPosition = 0
        listHelper.list.forEachIndexed { i, bean ->
            val itemAdapter =
                adaptersManager.getAdapter(bean.getBindAdapterClass()).castSuperAdapter()
            val itemCount = itemAdapter.getItemCount(bean)
            val nextStartPosition = itemStartPosition + itemCount

            if (nextStartPosition > absListPosition) {
                //下一个adapter的位置比position大说明当前type就在这个adapter中

                val itemPosition = absListPosition - itemStartPosition

                //当前状态
                val isFirst = absListPosition == 0
                val isLast = absListPosition == listHelper.list.lastIndex

                if (internalRecycle) {
                    //内部使用则复用单独一个对象
                    val info = internalLastCachePositionInfo?.also {
                        it.absListPosition = i
                        it.itemRelativePosition = itemPosition
                        it.itemAdapter = itemAdapter
                        it.hasHeader = isHeaderView
                        it.hasFooter = isFooterView
                        it.isFirst = isFirst
                        it.isLast = isLast
                    } ?: ItemAdapterPositionInfo(
                        absListPosition = i,
                        itemPosition = itemPosition,
                        itemAdapter = itemAdapter,
                        hasHeader = isHeaderView,
                        hasFooter = isFooterView,
                        isFirst = isFirst,
                        isLast = isLast
                    )
                    internalLastCachePositionInfo = info
                    return info
                }
                val info = ItemAdapterPositionInfo(
                    absListPosition = i,
                    itemPosition = itemPosition,
                    itemAdapter = itemAdapter,
                    hasHeader = isHeaderView,
                    hasFooter = isFooterView,
                    isFirst = isFirst,
                    isLast = isLast
                )
                lastCachePositionInfo = info
                return info
            } else {
                //循环相加
                itemStartPosition = nextStartPosition
            }
        }
        throw RuntimeException("没有取到对应的type,可能你没有(及时)刷新adapter")
    }

    /**
     * position、adapter、class唯一并且可以互相取值
     */
    private inner class MyAdaptersManager {
        val map = SimpleArrayMap<Class<out BaseContainerItemAdapter<*>>, Int>(8)
        val list = ArrayList<BaseContainerItemAdapter<*>>(8)

        fun addAdapter(adapters: List<BaseContainerItemAdapter<*>>) {
            adapters.forEach { adapter ->
                adapter.attachContainer(this@BaseContainerAdapter)
                adapter.registerDataSetObserver(childObservers)
                if (!map.containsKey(adapter.javaClass)) {
                    list.add(adapter)
                    map.put(adapter.javaClass, list.lastIndex)
                }
            }
        }

        fun getAdapter(position: Int): BaseContainerItemAdapter<*> {
            return list.getOrNull(position)
                ?: throw RuntimeException("缺少对应的adapter，adapter数量：" + list.size + "，当前index：" + position)
        }

        fun getAdapter(cls: Class<out BaseContainerItemAdapter<*>?>): BaseContainerItemAdapter<*> {
            val index = map[cls] ?: throw RuntimeException("缺少对应的adapter：$cls")
            return list[index]
        }

        fun getPosition(cls: Class<out BaseContainerItemAdapter<*>>): Int {
            return map[cls] ?: throw NullPointerException("一般是数据变化没有(及时)刷新adapter导致的")
        }

        fun remove(position: Int) {
            list.removeAt(position).also { map.remove(it.javaClass) }
        }

        fun remove(cls: Class<out BaseContainerItemAdapter<*>>?) {
            map.remove(cls)?.also { list.removeAt(it) }
        }

        fun clear() {
            list.clear()
            map.clear()
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    // 以下是可调用方法
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    /**
     * 添加adapter.重复则不会被添加,必须先删除
     * 当然可以预先添加用不到的adapter
     */
    fun addAdapter(vararg adapters: BaseContainerItemAdapter<*>) {
        addAdapter(adapters.toList())
    }

    /**
     * 添加adapter.重复则不会被添加,必须先删除
     * 当然可以预先添加用不到的adapter
     */
    fun addAdapter(adapters: List<BaseContainerItemAdapter<*>>) {
        adaptersManager.addAdapter(adapters)
        checkLayoutManager()
        notifyDataSetChanged()
    }

    /**
     * 删除指定adapter
     *
     * @param adapterPosition 按添加顺序第几个
     */
    fun removeAdapter(adapterPosition: Int) {
        adaptersManager.remove(adapterPosition)
        notifyDataSetChanged()
    }

    /**
     * 删除指定adapter
     *
     * @param adapterClass 哪个adapter
     */
    fun removeAdapter(adapterClass: Class<out BaseContainerItemAdapter<*>>?) {
        adaptersManager.remove(adapterClass)
        notifyDataSetChanged()
    }

    /**
     * 清空adapter
     */
    fun removeAllAdapter() {
        adaptersManager.clear()
        notifyDataSetChanged()
    }

    /**
     * 返回所有的adapter
     */
    fun getAdapters(): List<BaseContainerItemAdapter<*>> {
        //为了安全起见，不允许私自增删
        return ArrayList(adaptersManager.list)
    }

    override val list = listHelper.list

    private fun getSuggestDefClickListener() =
        onItemClickListener as? DefOnItemClickListener<BEAN> ?: DefOnItemClickListener()

    /**
     * 建议使用[setOnItemClickListener]、[setOnItemLongClickListener]
     *
     * 自定义点击效果，包括点击长按等，需要熟悉[OnItemClickListener]类
     * 这个回调和子adapter的事件回调都会被调用（这里先调，子adapter后调）
     *
     * 注意同样逻辑别写重复了
     * 如果没有收到回调请检查你的adapter有没有对[RecyclerView.ViewHolder.itemView]设置了点击事件
     *
     * position为相对position，想获得绝对位置[BaseViewHolder.commonPosition]
     */
    override fun setOnItemClickListener(listener: OnItemClickListener<BEAN>?) {
        onItemClickListener = listener
    }

    /**
     * 点击回调
     *
     * 注意事项同上
     */
    fun setOnItemClickListener(
        clickListener: ((
            view: View,
            relativePosition: Int,
            currentBean: BEAN,
            vh: BaseViewHolder<*>,
            itemAdapter: BaseContainerItemAdapter<*>,
            containerAdapter: BaseContainerAdapter<*>
        ) -> Unit)?
    ) {
        setOnItemClickListener(getSuggestDefClickListener().apply { onItemClick = clickListener })
    }

    /**
     * 长按回调
     *
     * 注意事项同上
     */
    fun setOnItemLongClickListener(
        longClickListener: ((
            view: View,
            relativePosition: Int,
            currentBean: BEAN,
            vh: BaseViewHolder<*>,
            itemAdapter: BaseContainerItemAdapter<*>,
            containerAdapter: BaseContainerAdapter<*>
        ) -> Boolean)?
    ) {
        setOnItemClickListener(getSuggestDefClickListener().apply {
            onItemLongClick = longClickListener
        })
    }

    /**
     * adapter里的view设置tag后（使用[addItemViewClickWithTag]、[dispatchItemViewClickWithTag]）点击会回调此方法
     */
    fun setOnItemViewClickListenerWithTag(
        clickListener: ((
            view: View,
            relativePosition: Int,
            currentBean: BEAN,
            vh: BaseViewHolder<*>,
            itemAdapter: BaseContainerItemAdapter<*>,
            containerAdapter: BaseContainerAdapter<*>,
            tag: String
        ) -> Unit)?
    ) {
        setOnItemClickListener(getSuggestDefClickListener().apply {
            onItemViewClickWithTag = clickListener
        })
    }

    /**
     * adapter里的view设置tag后（使用[addItemViewClickWithTag]、[dispatchItemViewLongClickWithTag]）长按会回调此方法
     */
    fun setOnItemViewLongClickListenerWithTag(
        longClickListener: ((
            view: View,
            relativePosition: Int,
            currentBean: BEAN,
            vh: BaseViewHolder<*>,
            itemAdapter: BaseContainerItemAdapter<*>,
            containerAdapter: BaseContainerAdapter<*>,
            tag: String
        ) -> Boolean)?
    ) {
        setOnItemClickListener(getSuggestDefClickListener().apply {
            onItemViewLongClickWithTag = longClickListener
        })
    }

    override fun getOnItemClickListener() = onItemClickListener

    /**
     * 根据bean对象和adapter的相对位置获取绝对位置
     *
     * @param relativePosition 相对potion
     */
    fun getAbsPosition(bean: IContainerBean, relativePosition: Int): Int {
        var position = relativePosition
        listHelper.list.forEach { listBean ->
            if (listBean === bean) {
                return position + headerViewCount
            } else {
                val itemAdapter =
                    adaptersManager.getAdapter(bean.getBindAdapterClass()).castSuperAdapter()
                position += itemAdapter.getItemCount(bean)
            }
        }
        throw RuntimeException("在list中没有找到传入的bean对象$bean")
    }

    /**
     * 根据绝对position获取对应adapter的额外信息
     *
     * @param absPosition 一般为[BaseViewHolder.commonPosition]
     * @return 不建议声明为final，因为[notifyItemChanged]相关方法时并不会更新里面的position
     */
    @MainThread
    fun getItemAdapterPositionInfo(absPosition: Int): ItemAdapterPositionInfo {
        return getCacheItemPositionInfo(absPosition, false)
    }

    /**
     * 根据bean和相对position获取对应adapter的额外信息
     *
     * @param relativePosition 相对potion
     * @return 不建议声明为final，因为[notifyItemChanged]相关方法时并不会更新里面的position
     */
    @MainThread
    fun getItemAdapterPositionInfo(
        bean: IContainerBean,
        relativePosition: Int
    ): ItemAdapterPositionInfo {
        return getItemAdapterPositionInfo(getAbsPosition(bean, relativePosition))
    }

    /**
     * 把rv的LayoutManager改成其他的GridLayoutManager时，此方法理论上没啥用
     */
    fun changedLayoutManager(manager: GridLayoutManager) {
        lastLayoutManager = manager
        manager.spanSizeLookup = object : SpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                if (isHeaderView && position == 0) {
                    return manager.spanCount
                } else if (isFooterView && itemCount == position + 1) {
                    return manager.spanCount
                }
                val info = getCacheItemPositionInfo(position, true)
                val itemAdapter = info.itemAdapter.castSuperAdapter()
                return itemAdapter.getSpanSize(
                    list[info.absListPosition],
                    info.itemRelativePosition
                )
            }
        }
    }

    override var headerView: View?
        get() = listHelper.headerView
        set(value) {
            listHelper.headerView = value
        }
    override var footerView: View?
        get() = listHelper.footerView
        set(value) {
            listHelper.footerView = value
        }
}